### ISSUE TITLE
Infinite loop in SemaHLSL

### ISSUE THREAD

**Issue by ehsannas** [Microsoft/DirectXShaderCompiler#781]
The following HLSL code causes an infinite loop in SemaHLSL.cpp

```hlsl
struct KillerStruct {};

struct InnerStruct {
  KillerStruct s;
};

struct OuterStruct {
  InnerStruct s;
};

cbuffer Params_cbuffer : register(b0) {
  OuterStruct constants;
};

float4 main(float4 pos : POSITION) : SV_POSITION { return float4(0, 0, 0, 0); }
```

The reason seems to be because the `if` condition below does not catch an empty struct, and there is infinite recursive calls to `IsTypeNumeric`. I think the proper fix might be in the flattened iterator rather than this function. So someone who is more familiar with this code should take a look.

```cpp
bool HLSLExternalSource::IsTypeNumeric(QualType type, UINT* count)
{
...
  case AR_TOBJ_COMPOUND:
    {
      FlattenedTypeIterator it(SourceLocation(), type, *this);
      // Return false for empty struct.
      if (!it.hasCurrentElement())
        return false;
      while (it.hasCurrentElement()) {
        bool isFieldNumeric = IsTypeNumeric(it.getCurrentElement(), &subCount);
...
}
```
Thanks

**Comment by ehsannas** [comment]
cc @chaoticbob

**Comment by python3kgae** [comment]
Fixed with https://github.com/Microsoft/DirectXShaderCompiler/pull/760

**Comment by ehsannas** [comment]
Thanks @python3kgae
